File: frontend/src/types/story.ts
export interface PaginatedResponse<T> {
  content: T[];
  pageable: any;
  totalPages: number;
  totalElements: number;
  last: boolean;
  size: number;
  number: number;
  sort: any;
  numberOfElements: number;
  first: boolean;
  empty: boolean;
}

export interface Story {
  id: number;
  title: string;
  slug: string;
  description: string;
  authorId: number;
  authorName: string;
  coverImageUrl: string | null;
  authorAvatarUrl?: string | null;
  status: string;
  wordCount: number;
  viewCount: number;
  followCount: number;
  favoriteCount: number;
  commentCount: number;
  chapterCount: number;
  createdAt: string;
  updatedAt: string;
  latestChapterNumber?: number | null;
  latestChapterTitle?: string | null;
  publishedChapterCount?: number;
  draftChapterCount?: number;
  categories?: { id: number; name: string; slug: string; description: string }[];
  tags?: { id: number; name: string; slug: string }[];
  visibility?: string;
  ageRating?: string | null;
  contentWarning?: string | null;
}

export interface TocChapter {
  id: number;
  chapterNumber: number;
  slug: string;
  title: string;
  publishedAt: string | null;
  wordCount: number;
}

export interface PublicStoryDetail extends Story {
  authorUsername: string;
  categories: any[];
  tags: any[];
  chapters: TocChapter[];
}

export interface ReadChapterResponse {
  id: number;
  storyId: number;
  title: string;
  chapterNumber: number;
  slug: string;
  content: string;
  wordCount: number;
  estimatedReadingTime: number;
  publishedAt: string | null;
  prevChapterSlug: string | null;
  nextChapterSlug: string | null;
  viewCount: number;
  likeCount: number;
  commentCount: number;
  hasLiked: boolean;
  lastReadPosition?: number;
}

File: backend/src/main/java/com/abora/backend/story/Story.java
package com.abora.backend.story;

import com.abora.backend.chapter.Chapter;
import com.abora.backend.user.User;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;
import lombok.NoArgsConstructor;
import lombok.AllArgsConstructor;
import lombok.Builder;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;

import java.time.Instant;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

@Entity
@Table(name = "stories")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Story {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "author_id", nullable = false)
    private User author;

    @Column(nullable = false, length = 255)
    private String title;

    @Column(nullable = false, unique = true, length = 255)
    private String slug;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Column(name = "cover_image_url", length = 500)
    private String coverImageUrl;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 30)
    @Builder.Default
    private StoryStatus status = StoryStatus.DRAFT;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 30)
    @Builder.Default
    private StoryVisibility visibility = StoryVisibility.PUBLIC;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "content_warning", columnDefinition = "json")
    private String contentWarning;

    @Column(name = "age_rating", length = 30)
    private String ageRating;

    @Column(nullable = false, length = 30)
    @Builder.Default
    private String language = "vi";

    @Column(name = "word_count", nullable = false)
    @Builder.Default
    private Integer wordCount = 0;

    @Column(name = "view_count", nullable = false)
    @Builder.Default
    private Long viewCount = 0L;

    @Column(name = "follow_count", nullable = false)
    @Builder.Default
    private Integer followCount = 0;

    @Column(name = "favorite_count", nullable = false)
    @Builder.Default
    private Integer favoriteCount = 0;

    @Column(name = "comment_count", nullable = false)
    @Builder.Default
    private Integer commentCount = 0;

    @Column(name = "chapter_count", nullable = false)
    @Builder.Default
    private Integer chapterCount = 0;

    @OneToMany(mappedBy = "story", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    @Builder.Default
    private List<Chapter> chapters = new ArrayList<>();

    @Column(name = "created_at", nullable = false, insertable = false, updatable = false)
    private Instant createdAt;

    @Column(name = "updated_at", nullable = false, insertable = false, updatable = false)
    private Instant updatedAt;

    @ManyToMany
    @JoinTable(
            name = "story_categories",
            joinColumns = @JoinColumn(name = "story_id"),
            inverseJoinColumns = @JoinColumn(name = "category_id")
    )
    @Builder.Default
    private Set<Category> categories = new HashSet<>();

    @ManyToMany
    @JoinTable(
            name = "story_tags",
            joinColumns = @JoinColumn(name = "story_id"),
            inverseJoinColumns = @JoinColumn(name = "tag_id")
    )
    @Builder.Default
    private Set<Tag> tags = new HashSet<>();
}
